### DISCUSSION TITLE
How to implement merge and flatten operation

### DISCUSSION THREAD

**Discussion by maxxxxwell** [gdsfactory/gdsfactory#3306] [Q&A]
Dear all
1:There are many shapes in the component such as s1 s2 s3 s4……sn(as an example I added 5 sturctures, but actually there are a large number of structure)，I want to perform the boolean(or) on  all of them，how to obtan all the elements in a component then performming boolean(or), is there a convinent and quick way? or it can be referred to a merge operation.
2:There are many sub cell，I want to remove all elements into one cell. in other words performing flatten operation.
3:There are "merge"and "flatten" button in klayout, but i want to implement them in gdsfactory by code. 

```python
import gdsfactory as gf

Layer1 = [1,0]

c = gf.Component('test')

s1 = c << gf.components.rectangle(size=(10,10),layer=Layer1)
s2 = c << gf.components.rectangle(size=(10,1),layer=Layer1)
s2.dmove([5,5])
s3 = c << gf.components.rectangle(size=(10,2),layer=Layer1)
s3.dmove([5,8])

w1 = gf.components.ellipse(radii=(3,3),layer = Layer1)
w2 = gf.components.ellipse(radii=(5,2),layer = Layer1)
s4 = gf.boolean(w1, w2, operation='or')
c << s4

sn = c << gf.components.ring(radius=10,width=1,layer=Layer1)

c.show()
```
![merge and flatten](https://github.com/user-attachments/assets/1779509f-22b5-495e-ae06-d69009fd4925)

**Comment by joamatab** [comment]
1. https://gdsfactory.github.io/gdsfactory/notebooks/04_components_geometry.html#boolean
2 - 3.

```python

import gdsfactory as gf

Layer1 = [1,0]

c = gf.Component('test')

s1 = c << gf.components.rectangle(size=(10,10),layer=Layer1)
s2 = c << gf.components.rectangle(size=(10,1),layer=Layer1)
s2.dmove([5,5])
s3 = c << gf.components.rectangle(size=(10,2),layer=Layer1)
s3.dmove([5,8])

w1 = gf.components.ellipse(radii=(3,3),layer = Layer1)
w2 = gf.components.ellipse(radii=(5,2),layer = Layer1)
s4 = gf.boolean(w1, w2, operation='or')
c << s4

sn = c << gf.components.ring(radius=10,width=1,layer=Layer1)

c.flatten()
c.show()
```

**Reply by maxxxxwell** [reply]
Thank you very much
